Fix grid layout

sobo-odoo opened the pull request:
This PR contains improvements and fixes of the grid mode that was added in [1].
The linked design-themes PR can be found at [2].

[1]: https://github.com/odoo/odoo/pull/93144
[2]: https://github.com/odoo/design-themes/pull/592

task-2973198

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/99857).

sobo-odoo commented:
@qsm-odoo it's ready for review :) (if no new issue appears in the mean time but I don't think it will)
FYI: there is one commit by fix/improvement and there is also a design-themes PR related to the Masonry commit [1].

[1]: https://github.com/odoo/design-themes/pull/592

qsm-odoo reviewed (changes requested):
Courage :muscle: Again, that's a complicated feature to implement with all the exception and the drag and drop and everything :grimacing: That's maybe as difficult to review, I tried to make a quick review so you can start working on it tomorrow already. The main point is that we should remove all website mentions which are in web_editor :grimacing: 

Courage :muscle:

sobo-odoo commented on addons/web_editor/static/src/js/editor/snippets.options.js:
```diff
@@ -4960,7 +4960,8 @@ registry.layout_column = SnippetOptionWidget.extend({
             // Set the columns properties.
-            newColumnEl.classList.add('col-lg-6', 'g-col-lg-6', 'g-height-6');
+            newColumnEl.classList.add('col-lg-6', 'g-col-lg-6', 'g-height-6', 'o_grid_item_image');
+            newColumnEl.contentEditable = false;
```
That is because it was the image that was selected. Even in normal mode, you have to click next to the image to make the blinking cursor appear and so to not replace the image. Here, this line is just to prevent to write text inside the column (so to not have the cursor). However, it unfortunately doesn't prevent the image to be removed/replaced if this is what was selected. 
I tried to find a solution to make the image "unremovable" but I didn't find one :/ (for example by preventing the user to select the image with `user-select: none`. With that, the blue highlight doesn't show (only on Chrome though...) but the image is still removable :/)
Do you have any suggestion ?

Note 1: the image is highlighted in blue when it is selected so we have an indication of that.
Note 2: the problem in Firefox is that the text appears inside the row but outside of any column (in Chrome the text goes in the previous column). I didn't have any crash but this is still problematic because it is incorrectly placed in the grid.

Knowing all that, what is the best ? Not editing but having weird effects if we don't see the selection highlight, or allowing the edition but having the ugly overlay overflow effect there was before ? (If you have no suggestion of course)

qsm-odoo replied:
> Do you have any suggestion ?

Not really, typing text at wrong locations has always been a problem that the editor team will likely solve the right way at some point. Meanwhile, maybe the class oe_unremovable (something like that) could prevent the image to be removed... but I think it would also prevent removing the image via the trash icon etc so... probably not great either :grimacing: I'll have to test better to understand/remember the real issue here but I am tempted to say "let's let people break stuff". As long as there is no traceback, letting people write at wrong places is not a big deal to me as they can always go back (normally) or type at the right places. But maybe I don't remember fully the issue that is being solved, we can talk about it next week.

sobo-odoo commented on addons/web_editor/static/src/js/common/grid_layout_utils.js:
```diff
@@ -179,6 +180,17 @@ function _placeColumns(columnEls, rowSize, rowGap, columnSize, columnGap) {
         columnSpans.push(columnSpan);
     }
 
+    // If we are in the Big Boxes snippet, set the padding of the grid items to
+    // the original padding of the boxes so they look like in normal mode.
+    if (isBigBoxes) {
```
Oh good idea I didn't think of that ! It even makes Banner look way better :o 
And so to answer your previous comment: with this, no it is not needed to detect `s_color_blocks_2` anymore.
Thanks :smile: 

(However, I think we really should find a good way to put padding on the columns individually :grimacing:)

qsm-odoo replied:
> (However, I think we really should find a good way to put padding on the columns individually grimacing)

I am sure grid mode will evolve over the next months / years indeed :+1:

sobo-odoo commented:
@qsm-odoo I made the changes with the higher priority, except for the images edition as you can see in the comment I left (feel free to remove the line with `contenteditable` if you conclude we shouldn't block the edition).
Sorry I couldn't move the website stuff in website on time... I will do that on Monday :/

qsm-odoo commented:
> I made the changes with the higher priority, except for the images edition as you can see in the comment I left (feel free to remove the line with `contenteditable` if you conclude we shouldn't block the edition). Sorry I couldn't move the website stuff in website on time... I will do that on Monday :/

No worries, I was not asking you stay until 22h+ to finish this but I appreciate you did :+1: I'll try to have a look at it this week-end and decide if I merge it as it is already or wait a bit more :+1:

sobo-odoo commented on addons/web_editor/static/src/js/editor/snippets.editor.js:
```diff
@@ -928,39 +928,59 @@ var SnippetEditor = Widget.extend({
 
+        // Snippets for which the grid mode should not be toggled on drag.
+        const untoggleableColumns = '.s_masonry_block, .s_showcase, .s_features_grid, .s_website_form, .s_media_list, .s_table_of_content, .s_process_steps';
+        const disableToggle = self.$target[0].closest(untoggleableColumns);
```
I could use that directly if there was not the special case of Masonry: the option excludes this snippet because we don't need the option to toggle because it will always be in grid mode. But since it is in grid mode, it should still be "toggleable" in `_dragAndDropStart` given the way I made the condition (it just won't toggle for real because I check if the `o_grid_mode` class is there before doing so). Moreover, it's not visible here because I added it in a later commit but now it is `.s_masonry_block:not([data-vxml])` to support the old version of Masonry (so it's not ugly when we drag it).

So instead, I defined a list in the website part (as you suggested above)

qsm-odoo replied:
> Moreover, it's not visible here because I added it in a later commit but now it is .s_masonry_block:not([data-vxml]) to support the old version of Masonry

That's fine, but in general: we don't support using options on snippet of older data-XXX version. The only purpose of the data-XXX system is to keep the snippets working *outside of edit mode* and indicate in edit mode those are outdated snippets (if someone drag and see something weird, it would be ok as you advertise in edit mode this is an outdated snippet that should be replaced).

> I could use that directly if there was not the special case of Masonry

Mmh, can't this be "if the grid option is displayed in the editor panel **or if the snippet is already in grid mode**" ? :thinking:

qsm-odoo replied:
Keep in mind that the community can create snippets to extend Odoo. That's why most of our UI definition is in XML the way it is. Having to fill a javascript list in a random code the community is not supposed to care about is not dev-friendly. Everything should be automatic when possible :+1:

sobo-odoo commented:
@qsm-odoo I made all the changes :)  
FYI: In addition to the corrections, I also made the following changes since the last reviews:
- I removed the use of `word-break` and use `min-width` instead

- I added the fact that at the `start` of `SnippetEditor`, the `contenteditable` attribute is set to false for the columns with the `o_grid_item_image` class (because it is removed at the destroy so they needed to be added back when entering the edit mode again)

- I added a missing `contenteditable` in the design-themes PR

- I don't do anything special for the old masonry anymore (since I check if the grid layout option is present or if it has the grid mode instead of using a list (as you suggested), it is automatically taken into account)

- I removed the temporary fix I made in [1] in the undo and redo functions. Since the fix I made for the history of the resizing has been merged, it was not needed anymore. (-> new commit)

- PDE asked me to change the fact that the grid item we are resizing/moving goes in front of all the others, so I changed that.  
(-> new commit)

[1]: https://github.com/odoo/odoo/pull/93144

qsm-odoo reviewed (approved):
@sobo-odoo Thanks for the work :+1: I tested, it seems nice. [I also push-forced ](https://github.com/odoo/odoo/compare/9761ef1bb2b380b442a9c3ec8ae4cd9ea68bb899..b669e6166632bbe11aaea67ab1469b0785b09cf0) to fix a few details:

- 2-3 commits were mentioning the wrong apps
- Removed your diff `data-drop-in="section:not(.s_carousel_wrapper) .content, nav"`: actually I always wondered what that "content" class was (I thought it was historic and we should remove it some day)... it's actually a bug in stable without grid mode: you can take an "alert" snippet and put it in a `.row` of the carousel which have the "content" class for some strange reason. We should fix that in stable or at least in a dedicated commit and investigate (in 16.0, could be nice to remove the rule about "content" classes entirely if it make sense)
- `contentEditable="false"` => `contenteditable="false"` (while the version with the uppercase should work on all browsers, the HTML spec says it should be `contenteditable`)
- Mention merged commits instead of merged PR in commit messages (although I only did it for one commit, I was too lazy to change the rest). Mentioning commits is better when you can as you can check them in local, while you cannot check PR in local (or at least not as easily).

While the current result is already really cool, I think there is still room for improvement after this PR (cc @pierre-pde, to discuss with him), here are some:

- Some masonry templates should be better (I did not compare to before, maybe they were not nice before), I noted:
    - "Image - Text - Image" -> weird first image
    - "Mosaic" -> it really just is ugly ^^ Maybe taller boxes for the one with texts?
    - "Alternate Text" -> Taller boxes ?
    - "Alternate Text Image" -> weird last image
    - "Alternate Image Text" -> weird too

- Big boxes (probably others, like Masonry too?): There is no background options on those snippet, which is what we wanted since the boxes have those options and the boxes cover the whole snippet... but when you enter grid mode, that's quickly not true anymore. Could stay like that though but worth discussing IMO.

- Masonry padding option: as discussed, I asked to have the option (because paddings are necessary) but we should review the fact it does nothing visible while there is not much text (maybe just add a preview of the padding (like the "Content Width" option does on hover) to indicate what the effect is)

- Add a "big boxes" snippet (or a "text - image" one but remove the bottom padding), switch to grid. Then add a "masonry" snippet below. Select the first box of the masonry. Start dragging... and you end up in the snippet big boxes above by mistake. This is because in that case, the drag and drop button is not inside the snippet when starting dragging.

Bug again, nothing critical :+1:  Let's merge !

@robodoo rebase-ff r+

robodoo commented:
Merge method set to rebase and fast-forward.
